### DISCUSSION TITLE
Lesson%201%3A%20Hello%20Metal/1.%20Hello%20Window/

### DISCUSSION THREAD

**Discussion by giscus** [wmarti/MetalTutorial#2] [Announcements]
# Lesson%201%3A%20Hello%20Metal/1.%20Hello%20Window/

https://metaltutorial.com/Lesson%201%3A%20Hello%20Metal/1.%20Hello%20Window/

**Comment by nnoujaim** [comment]
I really appreciate these tutorials, thanks for making them and looking forward to more! I have some questions about GLFW though, can you plug in an audio engine to it like OpenAL or alternatives? Im trying to make a game with GLFW and Metal and thinking about down the line if this will be sufficient for everything a game needs.

**Reply by wmarti** [reply]
That’s a great question. I’ve actually never done audio before, so my suggestion would be to check out something like [this](https://learnopengl.com/In-Practice/2D-Game/Audio). They use GLFW for windowing and mouse/keyboard input as well, but I don’t believe it supports audio. Alternatively, you could look into a different windowing system like SDL2, they might have audio capabilities built in. I'll consider doing a chapter on this in the future, thanks for the suggestion. 😃

**Reply by nnoujaim** [reply]
Np, thanks for the quick reply. I actually started out trying to use SDL because it's what tons of people on the internet use, and it apparently does have some type of low level audio functionality plus can integrate with some audio engines, but unfortunately there aren't any great straight forward tutorials that a beginner could understand like this one for C++ + SDL + Metal. You have a great teaching style!

The tutorials I have found are either for Swift + Metal, or assume that you understand lots of little things that experienced programmers would know but that make it pretty difficult to get started as a newcomer. I found some example github repo's too, but they either require you to patch different things together or there's custom header modifications done to Metal. I'd rather learn with the vanilla version of all of these libraries until I understand the fundamentals better.

Here's another suggestion - you should totally make an SDL version of this tutorial!

**Reply by wmarti** [reply]
Thank you for the kind comments, I'll definitely try to create more tutorials soon, although adapting the code into a lesson can be quite time consuming 😅. I'll definitely consider SDL for the future. For now, I would suggest asking ChatGPT for help with that, it may be able to assist better than I can, as I'm not very familiar with SDL.

**Comment by hanasiu** [comment]
After error, I checked 'Signing & Capabilities' -'disable Library Validation' and then it works. but I may not be fully following the tutorial, or it's possible that my environment is different.

**Reply by wmarti** [reply]
Hmm, did you download Lesson 1_1 directly from the GitHub? I'll try it out and update it if I experience the same issue.

**Reply by hanasiu** [reply]
I followed the tutorial without downloading any GitHub files. maybe my mistake. Thank you!

**Reply by zzhhdaaa** [reply]
Getting same error and fixed using the same method! It seems that checking this 'Disable Library Validation' will auto generate a 'xxx.entilements' file under the project dir.

**Reply by Ajaya1000** [reply]
Works if we change it to "Embed & sign"

**Comment by RW-77** [comment]
How do I create a .mm file? I don't see any option to do so.

**Reply by RW-77** [reply]
Create a normal .cpp file, right click on it, and show the File Inspector. In the name field, change the ending as needed. When viewing the file through the File Inspector afterwards, the file type should be Objective-C++.
